x/stakeibc/keeper/validator_selection_test.go

```diff
@@ -1,18 +1,159 @@
 package keeper_test
 
 import (
+	"fmt"
 	"math/rand"
 
 	sdkmath "cosmossdk.io/math"
+	"github.com/gogo/protobuf/proto" //nolint:staticcheck
 
 	sdk "github.com/cosmos/cosmos-sdk/types"
+	ibctesting "github.com/cosmos/ibc-go/v5/testing"
+
 	// TODO [LSM]: Revert type
 	lsmstakingtypes "github.com/iqlusioninc/liquidity-staking-module/x/staking/types"
 
+	epochtypes "github.com/Stride-Labs/stride/v9/x/epochs/types"
 	"github.com/Stride-Labs/stride/v9/x/stakeibc/keeper"
 	"github.com/Stride-Labs/stride/v9/x/stakeibc/types"
 )
 
+type RebalanceDelegationsForHostZoneTestCase struct {
+	expectedRebalancings []types.Rebalancing
+	channelStartSequence uint64
+	hostZone             types.HostZone
+	delegationChannelID  string
+	delegationPortID     string
+}
+
+func (s *KeeperTestSuite) SetupTestRebalanceDelegationsForHostZone() RebalanceDelegationsForHostZoneTestCase {
+	// TODO [LSM]: Fix after merge
+	// delegationChannelID, delegationPortID := s.CreateICAChannel("cosmos.DELEGATION")
+	delegationPortID := "icacontroller-GAIA.DELEGATION"
+	delegationAccountOwner := fmt.Sprintf("%s.%s", HostChainId, "DELEGATION")
+	delegationChannelID := s.CreateICAChannel(delegationAccountOwner)
+
+	// Add host zone and validators
+	delegationAddress := "cosmos_DELEGATION"
+	hostZone := types.HostZone{
+		ChainId:              HostChainId,
+		HostDenom:            Atom,
+		DelegationIcaAddress: delegationAddress,
+		ConnectionId:         ibctesting.FirstConnectionID,
+		Validators: []*types.Validator{
+			// Total delegation: 10000
+			{Address: "val1", Weight: 25, Delegation: sdkmath.NewInt(3500)}, // Expected: 2500
+			{Address: "val2", Weight: 50, Delegation: sdkmath.NewInt(2000)}, // Expected: 5000
+			{Address: "val3", Weight: 25, Delegation: sdkmath.NewInt(4500)}, // Expected: 2500
+		},
+	}
+	s.App.StakeibcKeeper.SetHostZone(s.Ctx, hostZone)
+
+	// Add the stride epoch to determine the ICA timeout
+	strideEpochTracker := types.EpochTracker{
+		EpochIdentifier:    epochtypes.STRIDE_EPOCH,
+		EpochNumber:        1,
+		NextEpochStartTime: uint64(s.Coordinator.CurrentTime.UnixNano() + 30_000_000_000), // dictates timeouts
+	}
+
+	s.App.StakeibcKeeper.SetEpochTracker(s.Ctx, strideEpochTracker)
+
+	// Build expected redelegation messages
+	expectedRebalancings := []types.Rebalancing{
```

Comment:
Here we only have two rebalancing messages but in reality we batch these when there are many,  Do you want to make a test that verifies the batched sends and their callbacks don't get messed up?

Reply:
yeah that's a good call

Reply:
done